rand_os/src/lib.rs

```diff
@@ -9,7 +9,112 @@
 
 //! Interface to the random number generator of the operating system.
 //!
-//! See `OsRng` documentation for more details.
+//! `OsRng` is the preferred external source of entropy for most applications.
+//! Commonly it is used to initialize a user-space RNG, which can then be used
+//! to generate random values with much less overhead than `OsRng`.
+//!
+//! You may prefer to use [`EntropyRng`] instead of `OsRng`. It is unlikely, but
+//! not entirely theoretical, for `OsRng` to fail. In such cases [`EntropyRng`]
+//! falls back on a good alternative entropy source.
+//!
+//! `OsRng::new()` is guaranteed to be very cheap (after the first successful
+//! call), and will never consume more than one file handle per process.
+//!
+//! # Usage example
+//! ```
+//! use rand_os::OsRng;
+//! use rand_os::rand_core::RngCore;
```

Comment:
Wait, you made this a public re-export? Is there any convention here?

I see it's convenient, and I suppose it's fairly obvious. *Probably* it also works in edition 2018, though otherwise that would let us skip the `extern crate` bit.

Reply:
Yes, I think it's a better approach than re-exporting all items or requiring an explicit dependency on `rand_core`. In RustCrypto in addition to re-exporting crate I re-export most often used traits, we could do it here as well, it will shorten imports to `use rand_os::{OsRng, RngCore};`.

Though I hope we will get [inherent trait impls](https://github.com/rust-lang/rfcs/pull/2375) one day.

Reply:
I think in 2018 edition we will be able to write `pub use rand_core;`, or re-export crate via `Cargo.toml` (IIRC it's how crate renaming will work).

Reply:
I wasn't aware that other RNG crates didn't re-export `rand_core`. I believe it's not a "feature", but an omission. From my RustCrypto experience requiring users to explicitly depend on a trait crate will be a significant annoyance for no good reason and will cause errors like "`ChachaRng does not implement `RngCore``" on versions mismatch, which are quite puzzling even for intermediate/advanced users. (I think @tarcieri will agree here :) )

So I believe as a minimum we should re-export `rand_core` in RNG crates, and discuss whether we want to additionally re-export `RngCore` and `SeedableRng`.

Such re-export may become unnecessary with inherent trait impls, but we are not there yet.

Reply:
I do agree I find this approach convenient in the RustCrypto crates, and have started doing it myself in similar situations in some of my crates.

Reply:
I agree completely with that error message being extremely unintuitive, but..  

I disagree that reexporting addresses the problem because:  You still encounter the same error message but now miss-versioning any reexporting crate causes the problem.  If crates do not reexport then you normally miss-versioned the trait's own crate and you detect version conflict after one `cargo check`, `vim ../Cargo.toml`, `cargo check` sequence.  It seem reexporting makes errors even less intuitive.

We need better error message for trait version miss-matches.  We should consider some method for crates to impose correct versioning.  It's conceivable reexports could play some role here, if for example `use chacharng::Rng; use rand::Rng;` worked but required both `Rng` traits be identical.  It's likely better to leverage some "inherent traits" mechanism, so if for example if we have 
```
#![inherent]
impl RngCore for ChaChaRng { }

#![inherent]
impl RngCore for OsRng { }
```
then we see an informative warning or error if both `ChaChaRng` and `OsRng` are used in the same scope with different `RngCore`s.

Reply:
We are only talking about version-mismatch of the `rand_core` crate here — for this crate we can probably always make a compatibility shim (like 0.2.2 is) which should solve most of these errors.

In Rust 2018 users don't need to add `extern rand_core;` any more, so adding an extra dependency only involves two lines (`rand_core = "0.3"` and `use rand_core::RngCore;`).

There is a third point that `RngCore` isn't the front-end trait aimed at users: that's `Rng`. (This may be irrelevant though, since anyone using `Rng` can just import `OsRng` from `rand`.)

Reply:
>so adding an extra dependency only involves two lines (rand_core = "0.3" and use rand_core::RngCore;)

Either way it's an annoyance for users and as I've said earlier for no good reason. Users still have to track `rand_core` version, and maintainers will have to remember to update its version in RNG crate docs. It's simply an extra-work, which does not bring anything to the table.

If users want only the basic RNG functionality why make their live harder?

Reply:
In this case, wouldn't it be preferable to re-export just the parts of `rand_core` that users might want, so that they can write `use rand_os::{OsRng, RngCore};`?

There are [three traits and two error-related types](https://docs.rs/rand_core/0.3.0/rand_core/) to re-export; the modules are only for implementing `RngCore` so shouldn't be re-exported.

This isn't just about convenience; since Cargo can't guarantee that `rand_os::rand_core` is the same crate as `rand_core` (assuming there is also a direct dependence), it is probably less confusing to re-export under that name?

Reply:
I think it should be enough to re-export `rand_core` and `RngCore`. I think having re-exported crate is useful for discoverability (i.e. users will need only one click to read `rand_core` docs) and you will have less items to re-export. And with inherent trait impls we will be able to remove `RngCore` re-export as well.

>since Cargo can't guarantee that rand_os::rand_core is the same crate as rand_core, it is probably less confusing to re-export under that name?

I am not sure if I understood you here. Generally `rand_core` can come from different places (e.g. from two RNG crates) and there is indeed a possibility for `rand_cor` version mismatch, but I think not much can be done here.